### ISSUE TITLE
Comments on the API

### ISSUE THREAD

**Issue by ulugbekna** [microsoft/vscode#226775]
Testing #226669

# Comments

1) `TextSearchProviderOptions#folderOptions` is an array; what's the expected behavior if the array is empty? 

2) `TextSearchProviderOptions#folderOptions` - all props within the literal object (eg `excludes, `useIgnoreFiles`) non-optional? Is that on purpose? 

3) `includes: string[]` but `excludes: GlobPattern[]` -- is that expected? 

4) 

> If set, {@link TextSearchProviderOptions.useIgnoreFiles.local} should also be `true`.

4.1. Would it make sense to model this with types? eg `{ local: false } | { local: true, parent: boolean; global: boolean }`
4.2. the `@link`s seem broken, but maybe it's not possible to ref such deep fields

5) How does includes/excludes work with followSymlinks? Should they be respected by files referenced by symlinks? 

6) `TextSearchProviderOptions#maxResults` seems non-optional. I wonder why. 

7) Again, `TextSearchProviderOptions#previewOptions#matchLines` is non-optional but says "defaults to 100"? 

8) `TextSearchCompleteNew#limitHit` could be `exceedsLimit` or something like this to be more expressive about "If exactly that number of matches exist, this should be false." 

9) I also don't quite understand if `includes/excludes` glob patterns apply with `folderOption#folder` as the root 

# Nits: 

1) 

> 		 * If explicitly contains a newline character (`\n`), the default search behavior
		 * will automatically enable {@link isMultiline}.
		 */

For clarity, I'd propose: 

```
- * If explicitly contains a newline character
+ * If pattern contains a newline character
```

2) does "default search behavior" mean "default search provider ? Seems like the only place where this phrasing is used

3) 

> 		 * `pattern` contains a newline character (`\n`).

maybe could use `{@link pattern}`

4) Maybe `TextSearchProviderOptions#folderOptions` big literal object type could be a named type because I wanted to process each `folderOption` in a util function but then realized that if I wanted to pass a folderOption as an argument, I would need to copy paste this gigantic type

**Comment by andreamah** [comment]
> 1. `TextSearchProviderOptions#folderOptions` is an array; what's the expected behavior if the array is empty?

It should never be empty, since the reason that the provider is called in the first place is because a folder is open with the relevant scheme.


> 2\. `TextSearchProviderOptions#folderOptions` - all props within the literal object (eg `excludes, `useIgnoreFiles`) non-optional? Is that on purpose?

Yes, it's given to the extension, so we try to make these non-optional as much as we can. Then, the extension knows all the info that it needs (instead of setting its own default values if info is missing).

> 3\. `includes: string[]` but `excludes: GlobPattern[]` -- is that expected?

Yes. the `folder` element is what the search is relative to. So all `includes` should be relative to `folder`, and it should not have a `baseUri` that is present in `GlobPattern`. This could be more clear, though- I'll add some docs for this.

> 4.1. Would it make sense to model this with types? eg `{ local: false } | { local: true, parent: boolean; global: boolean }`

See https://github.com/microsoft/vscode/issues/205692#issuecomment-2098828197. We didn't want to restrict the API shape too much, as our current logic is very ripgrep-based.

> 4.2. the `@link`s seem broken, but maybe it's not possible to ref such deep fields

🤔 Good catch. I'll see if I can figure this out

> How does includes/excludes work with followSymlinks? Should they be respected by files referenced by symlinks? 

The implication is that includes and excludes will respect symlinks. When you `useSymlinks`, there should probably be a symlink pathname that can be matched against for includes/excludes. I didn't state that in the desc for includes/excludes for brevity, but lmk if that seems like something I should add!

> 6\. `TextSearchProviderOptions#maxResults` seems non-optional. I wonder why.

We decided that maxResults should be defined all of the time. We don't want providers to go out-of-control with their match counts. If they want to set their own maxResults settings/defaults, they can easily ignore the setting (although not advised).


> 7\. Again, `TextSearchProviderOptions#previewOptions#matchLines` is non-optional but says "defaults to 100"?

I think I fixed this since then.


> 8\. `TextSearchCompleteNew#limitHit` could be `exceedsLimit` or something like this to be more expressive about "If exactly that number of matches exist, this should be false."

Hmm. We usually want providers to stop when the limit is hit. `exceedsLimit` implies that we continue past the limit, but just wanted to throw an fyi to the user.


> 9\. I also don't quite understand if `includes/excludes` glob patterns apply with `folderOption#folder` as the root

Yes. We only search from one folder at a time, with certain `includes` acting as patterns to include. Within this, `excludes` can be glob patterns that might have their own `baseUri` to act relative to. These will be avoided when searching in `folder` + `includes`.


Nits 1 & 3- yes, I can make that change.

Nit 2 - see https://github.com/microsoft/vscode/issues/226869

Nit 4 - https://github.com/microsoft/vscode/issues/226817

Thanks for the feedback!! One thing - it might be a little easier for TPI authors to respond to and address issues if they are broken up into different issues. Usually, I try to split them into different/related issues unless the TPI author states that they want them.
